### PR TITLE
bpo-1635741: Port _collections module to multiphase initialization. 

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue1635741

### CODE REVIEW

**Comment by corona10** [comment]
For the record, I will work on a new helper function after this PR is landed. 
https://github.com/python/cpython/pull/19071#discussion_r395074315

**Comment by corona10** [comment]
@shihai1991 Please take a look :)

File: Modules/_collectionsmodule.c [hunk 1]
```diff
@@ -2546,24 +2546,59 @@ static PyTypeObject tuplegetter_type = {
...
+    };
+
+    defdict_type.tp_base = &PyDict_Type;
+
+    for (int i = 0; typelist[i] != NULL; i++) {
+        PyTypeObject *type = typelist[i];
+        if (PyType_Ready(type) < 0) {
+            return -1;
+        }
+        const char *name = _PyType_Name(type);
```
Comment (vstinner) [inline Modules/_collectionsmodule.c:2579]: FYI I checked manually that the actual type name and _PyType_Name() gives the same name than before this change.

File: Modules/_collectionsmodule.c [hunk 2]
```diff
@@ -2546,24 +2546,59 @@ static PyTypeObject tuplegetter_type = {
...
+        Py_INCREF(type);
+        if (PyModule_AddObject(m, name, (PyObject *)type) < 0) {
```
Comment (vstinner) [inline Modules/_collectionsmodule.c]: Checking for PyModule_AddObject() error is a good thing ;-)

File: Modules/_collectionsmodule.c [hunk 3]
```diff
@@ -2546,24 +2546,59 @@ static PyTypeObject tuplegetter_type = {
...
 ");
 
-static struct PyMethodDef module_functions[] = {
+static struct PyMethodDef collections_methods[] = {
     _COLLECTIONS__COUNT_ELEMENTS_METHODDEF
     {NULL,       NULL}          /* sentinel */
 };
 
+static int
+collections_exec(PyObject *m) {
```
Comment (vstinner) [inline Modules/_collectionsmodule.c]: nitpick: would you mind to rename it to "mod" or "module"?

**Review by vstinner — APPROVED** [review]
LGTM apart "m" variable name which looks too short to me ;-)

File: Modules/_collectionsmodule.c (hunk 1 above)
Comment (corona10) [reply Modules/_collectionsmodule.c:2579]: I did same thing also :) Thank you for the re-checking!

File: Modules/_collectionsmodule.c (hunk 3 above)
Comment (corona10) [reply Modules/_collectionsmodule.c]: module might be great! Thanks for the suggestion

**Comment by shihai1991** [comment]
oh, on~ my network crashed at this point :( 
I can not review codes now, waiting a moment~

File: Modules/_collectionsmodule.c [hunk 4]
```diff
@@ -2546,24 +2546,59 @@ static PyTypeObject tuplegetter_type = {
...
+        &PyODict_Type,
+        &dequeiter_type,
+        &dequereviter_type,
+        &tuplegetter_type,
+        NULL,
+    };
+
+    defdict_type.tp_base = &PyDict_Type;
+
+    for (int i = 0; typelist[i] != NULL; i++) {
```
Comment (shihai1991) [inline Modules/_collectionsmodule.c:2574]: nice improvment.

**Review by shihai1991 — APPROVED** [review]
I like the improvment of typelist.
LGTM :)

**Comment by corona10** [comment]
@vstinner @shihai1991

Thanks for the reviews :)

**Pull request merged**